JSON: null value causes invalid memory address or nil pointer dereference error

Platform: Windows

example.json:
```
[{
  "nameCache": null
}]
```

settings.yaml
```
dataSources:
  - name: t
    type: json
    config:
      path: "example.json"
      arrayFormat: true
```

command
`octosql "select * from t as tt" -c settings.yaml`

result:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0xf8 pc=0x51bce6]

goroutine 1 [running]:
github.com/cube2222/octosql.NormalizeType(0x0, 0x0, 0xc00009b988, 0x2)
        /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}/values.go:235 +0xbb6
github.com/cube2222/octosql/storage/json.(*RecordStream).Next(0xc0001064b0, 0xc000079e60, 0x9c9e20, 0xc0001064b0)
        /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}/storage/json/datasource.go:120 +0x32d
github.com/cube2222/octosql/app.(*App).RunPlan(0xc00009bd18, 0x9cf420, 0xc0000700b0, 0x9c4740, 0xc0000de460, 0x0, 0xc00009bd38)
        /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}/app/app.go:45 +0x1a3
main.glob..func1(0xd31520, 0xc000079aa0, 0x1, 0x3)
        /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}/cmd/octosql/main.go:94 +0x594
github.com/spf13/cobra.(*Command).execute(0xd31520, 0xc00006c0d0, 0x3, 0x3, 0xd31520, 0xc00006c0d0)
        /go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:830 +0x2b5
github.com/spf13/cobra.(*Command).ExecuteC(0xd31520, 0xd36a90, 0x924ae8, 0x6)
        /go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x303
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
main.main()
        /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}/cmd/octosql/main.go:105 +0x157
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -2,12 +2,14 @@ workflows:
   version: 2
   test-release:
     jobs:
+      - lint
       - test
       - test-mysql
       - test-postgres
       - test-redis
       - release:
           requires:
+            - lint
             - test
             - test-mysql
             - test-postgres
@@ -15,6 +17,16 @@ workflows:
 
 version: 2
 jobs:
+  lint:
+    docker:
+      - image: circleci/golang:1.12
+
+    working_directory: /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}
+    steps:
+      - run: go get github.com/burntsushi/go-sumtype
+      - checkout
+      - run: go-sumtype ./...
+
   test:
     docker:
       - image: circleci/golang:1.12

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -21,7 +21,7 @@ jobs:
     docker:
       - image: circleci/golang:1.12
 
-    working_directory: /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}
+    working_directory: /go/src/github.com/cube2222/octosql
     steps:
       - run: go get github.com/burntsushi/go-sumtype
       - checkout
@@ -33,7 +33,7 @@ jobs:
     environment:
       GO111MODULE: "on"
 
-    working_directory: /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}
+    working_directory: /go/src/github.com/cube2222/octosql
     steps:
       - checkout
 
@@ -56,7 +56,7 @@ jobs:
           MYSQL_ROOT_PASSWORD: toor
           MYSQL_DATABASE: mydb
 
-    working_directory: /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}
+    working_directory: /go/src/github.com/cube2222/octosql
     steps:
       - checkout
 
@@ -74,7 +74,7 @@ jobs:
           POSTGRES_PASSWORD: toor
           POSTGRES_DB: mydb
 
-    working_directory: /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}
+    working_directory: /go/src/github.com/cube2222/octosql
     steps:
       - checkout
 
@@ -87,7 +87,7 @@ jobs:
           GO111MODULE: "on"
       - image: circleci/redis:latest
 
-    working_directory: /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}
+    working_directory: /go/src/github.com/cube2222/octosql
     steps:
       - checkout
 
@@ -99,7 +99,7 @@ jobs:
     environment:
       GO111MODULE: "on"
 
-    working_directory: /go/src/github.com/{{ORG_NAME}}/{{REPO_NAME}}
+    working_directory: /go/src/github.com/cube2222/octosql
     steps:
       - checkout
 

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -25,6 +25,7 @@ jobs:
     steps:
       - run: go get github.com/burntsushi/go-sumtype
       - checkout
+      - run: GO111MODULE=on go mod vendor
       - run: go-sumtype ./...
 
   test:

--- a/execution/functions/functions.go
+++ b/execution/functions/functions.go
@@ -62,10 +62,10 @@ var FuncInt = execution.Function{
 				return nil, err
 			}
 			return MakeInt(number), nil
-		default:
+		case Null, Phantom, Time, Duration, Tuple, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -103,10 +103,10 @@ var FuncFloat = execution.Function{
 				return nil, err
 			}
 			return MakeFloat(number), nil
-		default:
+		case Null, Phantom, Time, Duration, Tuple, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -131,10 +131,10 @@ var FuncNegate = execution.Function{
 			return -1 * arg, nil
 		case Float:
 			return -1 * arg, nil
-		default:
+		case Null, Phantom, Bool, String, Time, Duration, Tuple, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -165,10 +165,10 @@ var FuncAbs = execution.Function{
 				return -1 * arg, nil
 			}
 			return arg, nil
-		default:
+		case Null, Phantom, Bool, String, Time, Duration, Tuple, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -191,18 +191,18 @@ var FuncSqrt = execution.Function{
 		switch arg := args[0].(type) {
 		case Int:
 			if arg < 0 {
-				return nil, fmt.Errorf("Can't take square root of value %v", arg)
+				return nil, fmt.Errorf("can't take square root of value %v", arg)
 			}
 			return MakeFloat(math.Sqrt(float64(arg.AsInt()))), nil
 		case Float:
 			if arg < 0 {
-				return nil, fmt.Errorf("Can't take square root of value %v", arg)
+				return nil, fmt.Errorf("can't take square root of value %v", arg)
 			}
 			return MakeFloat(math.Sqrt(arg.AsFloat())), nil
-		default:
+		case Null, Phantom, Bool, String, Time, Duration, Tuple, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -227,10 +227,10 @@ var FuncFloor = execution.Function{
 			return MakeFloat(float64(arg.AsInt())), nil
 		case Float:
 			return MakeFloat(math.Floor(arg.AsFloat())), nil
-		default:
+		case Null, Phantom, Bool, String, Time, Duration, Tuple, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -255,10 +255,10 @@ var FuncCeil = execution.Function{
 			return MakeFloat(float64(arg.AsInt())), nil
 		case Float:
 			return MakeFloat(math.Ceil(arg.AsFloat())), nil
-		default:
+		case Null, Phantom, Bool, String, Time, Duration, Tuple, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -281,18 +281,18 @@ var FuncLog2 = execution.Function{
 		switch arg := args[0].(type) {
 		case Int:
 			if arg <= 0 {
-				return nil, fmt.Errorf("Can't take log of value %v", arg)
+				return nil, fmt.Errorf("can't take log of value %v", arg)
 			}
 			return MakeFloat(math.Log2(float64(arg.AsInt()))), nil
 		case Float:
 			if arg <= 0 {
-				return nil, fmt.Errorf("Can't take log of value %v", arg)
+				return nil, fmt.Errorf("can't take log of value %v", arg)
 			}
 			return MakeFloat(math.Log2(arg.AsFloat())), nil
-		default:
+		case Null, Phantom, Bool, String, Time, Duration, Tuple, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -315,18 +315,18 @@ var FuncLn = execution.Function{
 		switch arg := args[0].(type) {
 		case Int:
 			if arg <= 0 {
-				return nil, fmt.Errorf("Can't take ln of value %v", arg)
+				return nil, fmt.Errorf("can't take ln of value %v", arg)
 			}
 			return MakeFloat(math.Log1p(float64(arg.AsInt())) - 1), nil
 		case Float:
 			if arg <= 0 {
-				return nil, fmt.Errorf("Can't take ln of value %v", arg)
+				return nil, fmt.Errorf("can't take ln of value %v", arg)
 			}
 			return MakeFloat(math.Log1p(arg.AsFloat()) - 1), nil
-		default:
+		case Null, Phantom, Bool, String, Time, Duration, Tuple, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -360,10 +360,10 @@ var FuncLeast = execution.Function{
 			}
 
 			return MakeFloat(min), nil
-		default:
+		case Null, Phantom, Bool, String, Time, Duration, Tuple, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -396,10 +396,10 @@ var FuncGreatest = execution.Function{
 			}
 
 			return MakeFloat(max), nil
-		default:
+		case Null, Phantom, Bool, String, Time, Duration, Tuple, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -572,10 +572,10 @@ var FuncReverse = execution.Function{
 				out[len(out)-i-1] = el
 			}
 			return MakeTuple(out), nil
-		default:
+		case Null, Phantom, Int, Float, Bool, Time, Duration, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -622,7 +622,7 @@ var FuncMatchRegexp = execution.Function{
 	Logic: func(args ...Value) (Value, error) {
 		re, err := regexp.Compile(args[0].(String).AsString())
 		if err != nil {
-			return nil, fmt.Errorf("Couldn't compile regular expression")
+			return nil, fmt.Errorf("couldn't compile regular expression")
 		}
 
 		match := re.FindString(args[1].(String).AsString())
@@ -728,10 +728,10 @@ var FuncContains = execution.Function{
 				}
 			}
 			return MakeBool(false), nil
-		default:
+		case Null, Phantom, Int, Float, Bool, Time, Duration, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -768,10 +768,10 @@ var FuncIndex = execution.Function{
 				}
 			}
 			return MakeInt(-1), nil
-		default:
+		case Null, Phantom, Int, Float, Bool, Time, Duration, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -819,10 +819,10 @@ var FuncLength = execution.Function{
 			return MakeInt(len(arg)), nil
 		case Tuple:
 			return MakeInt(len(arg)), nil
-		default:
+		case Null, Phantom, Int, Float, Bool, Time, Duration, Object:
 			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-			panic("unreachable")
 		}
+		panic("unreachable")
 	},
 }
 
@@ -893,10 +893,10 @@ var FuncAdd = execution.Function{
 			switch args[0].(type) {
 			case Int, Float, Duration:
 				return args[0], nil
-			default:
+			case Null, Phantom, Bool, String, Time, Tuple, Object:
 				log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-				panic("unreachable")
 			}
+			panic("unreachable")
 
 		case 2:
 			switch arg := args[0].(type) {
@@ -908,10 +908,10 @@ var FuncAdd = execution.Function{
 				return arg + args[1].(Duration), nil
 			case Time:
 				return MakeTime(arg.AsTime().Add(args[1].(Duration).AsDuration())), nil
-			default:
+			case Null, Phantom, Bool, String, Tuple, Object:
 				log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-				panic("unreachable")
 			}
+			panic("unreachable")
 
 		default:
 			log.Fatalf("unexpected argument count in function: %v", len(args))
@@ -952,10 +952,10 @@ var FuncSubtract = execution.Function{
 				return arg * -1, nil
 			case Duration:
 				return arg * -1, nil
-			default:
+			case Null, Phantom, Bool, String, Time, Tuple, Object:
 				log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-				panic("unreachable")
 			}
+			panic("unreachable")
 
 		case 2:
 			switch arg := args[0].(type) {
@@ -967,10 +967,10 @@ var FuncSubtract = execution.Function{
 				return arg - args[1].(Duration), nil
 			case Time:
 				return MakeTime(arg.AsTime().Add(-1 * args[1].(Duration).AsDuration())), nil
-			default:
+			case Null, Phantom, Bool, String, Tuple, Object:
 				log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
-				panic("unreachable")
 			}
+			panic("unreachable")
 
 		default:
 			log.Fatalf("unexpected argument count in function: %v", len(args))
@@ -1012,14 +1012,14 @@ var FuncMultiply = execution.Function{
 				return MakeDuration(arg.AsDuration() * time.Duration(arg2.AsInt())), nil
 			case Float:
 				return MakeDuration(time.Duration(float64(arg.AsDuration()) * arg2.AsFloat())), nil
-			default:
+			case Null, Phantom, Bool, String, Time, Duration, Tuple, Object:
 				log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[1]).String())
-				panic("unreachable")
 			}
-		default:
-			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
 			panic("unreachable")
+		case Null, Phantom, Bool, String, Time, Tuple, Object:
+			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
 		}
+		panic("unreachable")
 	},
 }
 
@@ -1071,15 +1071,15 @@ var FuncDivide = execution.Function{
 			case Duration:
 				return MakeFloat(float64(arg.AsDuration()) / float64(arg2.AsDuration())), nil
 
-			default:
+			case Null, Phantom, Bool, String, Time, Tuple, Object:
 				log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[1]).String())
-				panic("unreachable")
 			}
-
-		default:
-			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
 			panic("unreachable")
+
+		case Null, Phantom, Bool, String, Time, Tuple, Object:
+			log.Fatalf("unexpected type in function: %v", reflect.TypeOf(args[0]).String())
 		}
+		panic("unreachable")
 	},
 }
 
@@ -1153,6 +1153,28 @@ var FuncDuration = execution.Function{
 			log.Fatalf("unexpected argument count in function: %v", len(args))
 			panic("unreachable")
 		}
+	},
+}
+
+var FuncImpute = execution.Function{
+	Name: "impute",
+	ArgumentNames: [][]string{
+		{"replacement", "maybe_null"},
+	},
+	Description: docs.List(
+		docs.Text("Return maybe_null if it's not null, otherwise return the replacement."),
+	),
+	Validator: All(
+		ExactlyNArgs(2),
+	),
+	Logic: func(args ...Value) (Value, error) {
+		switch arg := args[1].(type) {
+		case Null:
+			return args[0], nil
+		case Phantom, Int, Float, Bool, String, Time, Duration, Tuple, Object:
+			return arg, nil
+		}
+		panic("unreachable")
 	},
 }
 

--- a/execution/functions/functions.go
+++ b/execution/functions/functions.go
@@ -1156,25 +1156,49 @@ var FuncDuration = execution.Function{
 	},
 }
 
-var FuncImpute = execution.Function{
-	Name: "impute",
+var FuncCoalesce = execution.Function{
+	Name: "coalesce",
 	ArgumentNames: [][]string{
-		{"replacement", "maybe_null"},
+		{"...args"},
 	},
 	Description: docs.List(
-		docs.Text("Return maybe_null if it's not null, otherwise return the replacement."),
+		docs.Text("Returns the first non-null argument, or null if there isn't any."),
+	),
+	Validator: All(
+		AtLeastNArgs(1),
+	),
+	Logic: func(args ...Value) (Value, error) {
+		for i := range args {
+			switch arg := args[i].(type) {
+			case Null:
+				continue
+			case Phantom, Int, Float, Bool, String, Time, Duration, Tuple, Object:
+				return arg, nil
+			}
+			panic("unreachable")
+		}
+
+		return MakeNull(), nil
+	},
+}
+
+var FuncNullIf = execution.Function{
+	Name: "nullif",
+	ArgumentNames: [][]string{
+		{"to_replace", "target"},
+	},
+	Description: docs.List(
+		docs.Text("Returns null if target equals to_replace, returns target otherwise."),
 	),
 	Validator: All(
 		ExactlyNArgs(2),
 	),
 	Logic: func(args ...Value) (Value, error) {
-		switch arg := args[1].(type) {
-		case Null:
-			return args[0], nil
-		case Phantom, Int, Float, Bool, String, Time, Duration, Tuple, Object:
-			return arg, nil
+		if AreEqual(args[0], args[1]) {
+			return MakeNull(), nil
 		}
-		panic("unreachable")
+
+		return args[1], nil
 	},
 }
 

--- a/execution/functions/functions_test.go
+++ b/execution/functions/functions_test.go
@@ -879,39 +879,138 @@ func Test_various(t *testing.T) {
 			wantErr: false,
 		},
 		{
-			name: "impute(2, 5)",
+			name: "coalesce(2, 5)",
 			args: args{
 				args: []Value{
 					MakeInt(2),
 					MakeInt(5),
 				},
-				fun: FuncImpute,
-			},
-			want:    MakeInt(5),
-			wantErr: false,
-		},
-		{
-			name: "impute(2, <null>)",
-			args: args{
-				args: []Value{
-					MakeInt(2),
-					MakeNull(),
-				},
-				fun: FuncImpute,
+				fun: FuncCoalesce,
 			},
 			want:    MakeInt(2),
 			wantErr: false,
 		},
 		{
-			name: "impute('text', <null>)",
+			name: "coalesce(2, <null>)",
+			args: args{
+				args: []Value{
+					MakeInt(2),
+					MakeNull(),
+				},
+				fun: FuncCoalesce,
+			},
+			want:    MakeInt(2),
+			wantErr: false,
+		},
+		{
+			name: "coalesce('text', <null>)",
 			args: args{
 				args: []Value{
 					MakeString("text"),
 					MakeNull(),
 				},
-				fun: FuncImpute,
+				fun: FuncCoalesce,
 			},
 			want:    MakeString("text"),
+			wantErr: false,
+		},
+		{
+			name: "coalesce(<null>, 2)",
+			args: args{
+				args: []Value{
+					MakeNull(),
+					MakeInt(2),
+				},
+				fun: FuncCoalesce,
+			},
+			want:    MakeInt(2),
+			wantErr: false,
+		},
+		{
+			name: "coalesce(<null>, 'text')",
+			args: args{
+				args: []Value{
+					MakeNull(),
+					MakeString("text"),
+				},
+				fun: FuncCoalesce,
+			},
+			want:    MakeString("text"),
+			wantErr: false,
+		},
+		{
+			name: "coalesce(<null>, <null>, 1, 2)",
+			args: args{
+				args: []Value{
+					MakeNull(),
+					MakeNull(),
+					MakeInt(1),
+					MakeInt(2),
+				},
+				fun: FuncCoalesce,
+			},
+			want:    MakeInt(1),
+			wantErr: false,
+		},
+		{
+			name: "coalesce(<null>, <null>, <null>)",
+			args: args{
+				args: []Value{
+					MakeNull(),
+					MakeNull(),
+					MakeNull(),
+				},
+				fun: FuncCoalesce,
+			},
+			want:    MakeNull(),
+			wantErr: false,
+		},
+		{
+			name: "nullif('test', 'test')",
+			args: args{
+				args: []Value{
+					MakeString("test"),
+					MakeString("test"),
+				},
+				fun: FuncNullIf,
+			},
+			want:    MakeNull(),
+			wantErr: false,
+		},
+		{
+			name: "nullif('test', 'test2')",
+			args: args{
+				args: []Value{
+					MakeString("test"),
+					MakeString("test2"),
+				},
+				fun: FuncNullIf,
+			},
+			want:    MakeString("test2"),
+			wantErr: false,
+		},
+		{
+			name: "nullif('test', 5)",
+			args: args{
+				args: []Value{
+					MakeString("test"),
+					MakeInt(5),
+				},
+				fun: FuncNullIf,
+			},
+			want:    MakeInt(5),
+			wantErr: false,
+		},
+		{
+			name: "nullif(5, 5)",
+			args: args{
+				args: []Value{
+					MakeInt(5),
+					MakeInt(5),
+				},
+				fun: FuncNullIf,
+			},
+			want:    MakeNull(),
 			wantErr: false,
 		},
 	}

--- a/execution/functions/functions_test.go
+++ b/execution/functions/functions_test.go
@@ -878,6 +878,42 @@ func Test_various(t *testing.T) {
 			want:    MakeDuration(time.Second * 2),
 			wantErr: false,
 		},
+		{
+			name: "impute(2, 5)",
+			args: args{
+				args: []Value{
+					MakeInt(2),
+					MakeInt(5),
+				},
+				fun: FuncImpute,
+			},
+			want:    MakeInt(5),
+			wantErr: false,
+		},
+		{
+			name: "impute(2, <null>)",
+			args: args{
+				args: []Value{
+					MakeInt(2),
+					MakeNull(),
+				},
+				fun: FuncImpute,
+			},
+			want:    MakeInt(2),
+			wantErr: false,
+		},
+		{
+			name: "impute('text', <null>)",
+			args: args{
+				args: []Value{
+					MakeString("text"),
+					MakeNull(),
+				},
+				fun: FuncImpute,
+			},
+			want:    MakeString("text"),
+			wantErr: false,
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

--- a/execution/functions/table.go
+++ b/execution/functions/table.go
@@ -44,7 +44,8 @@ var usableFunctions = []execution.Function{
 	FuncMultiply,
 	FuncDivide,
 	FuncDuration,
-	FuncImpute,
+	FuncCoalesce,
+	FuncNullIf,
 }
 
 func init() {

--- a/execution/functions/table.go
+++ b/execution/functions/table.go
@@ -44,6 +44,7 @@ var usableFunctions = []execution.Function{
 	FuncMultiply,
 	FuncDivide,
 	FuncDuration,
+	FuncImpute,
 }
 
 func init() {

--- a/execution/functions/validators.go
+++ b/execution/functions/validators.go
@@ -216,6 +216,12 @@ func TypeOf(wantedType octosql.Value) *typeOf {
 
 func (v *typeOf) Validate(arg octosql.Value) error {
 	switch v.wantedType.(type) {
+	case octosql.Null:
+		if _, ok := arg.(octosql.Null); !ok {
+			return fmt.Errorf("expected type %v but got %v", reflect.TypeOf(octosql.ZeroNull()).String(), arg)
+		}
+		return nil
+
 	case octosql.Phantom:
 		if _, ok := arg.(octosql.Phantom); !ok {
 			return fmt.Errorf("expected type %v but got %v", reflect.TypeOf(octosql.ZeroPhantom()).String(), arg)

--- a/execution/order_by.go
+++ b/execution/order_by.go
@@ -41,9 +41,11 @@ func isSorteable(x octosql.Value) bool {
 		return true
 	case octosql.Time:
 		return true
-	default:
+	case octosql.Null, octosql.Phantom, octosql.Duration, octosql.Tuple, octosql.Object:
 		return false
 	}
+
+	panic("unreachable")
 }
 
 func compare(x, y octosql.Value) (int, error) {
@@ -113,9 +115,12 @@ func compare(x, y octosql.Value) (int, error) {
 		}
 
 		return 1, nil
-	default:
+
+	case octosql.Null, octosql.Phantom, octosql.Duration, octosql.Tuple, octosql.Object:
 		return 0, errors.Errorf("unsupported type in sorting")
 	}
+
+	panic("unreachable")
 }
 
 func (ob *OrderBy) Get(variables octosql.Variables) (RecordStream, error) {

--- a/execution/record.go
+++ b/execution/record.go
@@ -33,7 +33,7 @@ func (r *Record) Value(field octosql.VariableName) octosql.Value {
 			return r.data[i]
 		}
 	}
-	return nil
+	return octosql.MakeNull()
 }
 
 func (r *Record) Fields() []Field {

--- a/execution/relation.go
+++ b/execution/relation.go
@@ -96,11 +96,13 @@ func (rel *MoreThan) Apply(variables octosql.Variables, left, right Expression) 
 	case octosql.Time:
 		rightValue := rightValue.(octosql.Time)
 		return leftValue.AsTime().After(rightValue.AsTime()), nil
+	case octosql.Null, octosql.Phantom, octosql.Bool, octosql.Duration, octosql.Tuple, octosql.Object:
+		return false, errors.Errorf(
+			"invalid operands to more_than %v and %v with types %v and %v, only int, float, string and time allowed",
+			leftValue, rightValue, GetType(leftValue), GetType(rightValue))
 	}
 
-	return false, errors.Errorf(
-		"invalid operands to more_than %v and %v with types %v and %v, only int, float, string and time allowed",
-		leftValue, rightValue, GetType(leftValue), GetType(rightValue))
+	panic("unreachable")
 }
 
 type LessThan struct {

--- a/values.go
+++ b/values.go
@@ -11,10 +11,28 @@ import (
 	"github.com/cube2222/octosql/docs"
 )
 
+//go-sumtype:decl Value
 type Value interface {
 	docs.Documented
 	octoValue()
 	fmt.Stringer
+}
+
+type Null struct{}
+
+func (Null) octoValue()            {}
+func (v Null) AsNull() interface{} { return nil }
+func (v Null) String() string {
+	return "<null>"
+}
+func (v Null) Document() docs.Documentation {
+	return docs.Text("Null")
+}
+func MakeNull() Null {
+	return Null(struct{}{})
+}
+func ZeroNull() Null {
+	return struct{}{}
 }
 
 type Phantom struct{}
@@ -182,6 +200,8 @@ func ZeroObject() Object {
 // All types coming out of data sources have to be already normalized this way.
 func NormalizeType(value interface{}) Value {
 	switch value := value.(type) {
+	case nil:
+		return MakeNull()
 	case bool:
 		return MakeBool(value)
 	case int:
@@ -242,6 +262,20 @@ func AreEqual(left, right Value) bool {
 		return true
 	}
 	switch left := left.(type) {
+	case Null:
+		_, ok := right.(Null)
+		if !ok {
+			return false
+		}
+		return true
+
+	case Phantom:
+		_, ok := right.(Phantom)
+		if !ok {
+			return false
+		}
+		return true
+
 	case Int:
 		right, ok := right.(Int)
 		if !ok {

--- a/variables.go
+++ b/variables.go
@@ -46,7 +46,11 @@ func NewVariables(variables map[VariableName]Value) Variables {
 var ErrVariableNotFound = errors.New("variable not found")
 
 func (vs Variables) Get(k VariableName) (Value, error) {
-	return vs[k], nil
+	out, ok := vs[k]
+	if !ok {
+		return MakeNull(), nil
+	}
+	return out, nil
 }
 
 func (vs Variables) MergeWith(other Variables) (Variables, error) {

